### ISSUE TITLE
"Internal Server Error"? Where are detailed logs?

### ISSUE THREAD

**Issue by NetwarSystem** [google/timesketch#513]
This is Timesketch running on an Ubuntu 16.04 system with Elasticsearch 5.6.5

As soon as I click the plus sign for details with the intent to write something, it shows "Internet Server Error" at the top. I can write in the dialog box, but not save it.

https://imgur.com/a/QGhnW

I'm new to Timesketch, I've poked around a bit, other functions seem to be working, it's just this one.

**Comment by berggren** [comment]
Interesting. What format of the data are you importing? How did you import it, CSV import?

If you enable DEBUG in timesketch.conf you should see the backend error in the console.

**Comment by NetwarSystem** [comment]
I used the one line csv example from here and created the timeline via command line on the server.

Here is the log of the actual error. Note that I'm running Elastic 6.1.1 here, but saw similar trouble with 5.6.5.

https://gist.github.com/NetwarSystem/9e120d7b151f996bb86e2691071abcde

**Comment by NetwarSystem** [comment]
And reproduced on brand new install on a system running Elasticsearch 5.6.5, same data as above.

https://gist.github.com/NetwarSystem/c52ce88327e86634946bd45228ebbb6d

**Comment by NetwarSystem** [comment]
And the pip freeze output

https://gist.github.com/NetwarSystem/ca7f21ef34e76e6054c69474f69f1f83

**Comment by NetwarSystem** [comment]
here's the error:

  File "/usr/local/lib/python2.7/dist-packages/timesketch/api/v1/resources.py", line 786, in get
    result = self.datastore.get_event(searchindex_id, event_id)
  File "/usr/local/lib/python2.7/dist-packages/timesketch/lib/datastores/elastic.py", line 291, in get_event
    _source_exclude=[u'timesketch_label'])
  File "/usr/local/lib/python2.7/dist-packages/elasticsearch/client/utils.py", line 76, in _wrapped
    return func(*args, params=params, **kwargs)
TypeError: get() takes at least 4 arguments (4 given)

And here's the code in question - says get() needs four parms, I only count three here.

        try:
            # Suppress the lint error because elasticsearch-py adds parameters
            # to the function with a decorator and this makes pylint sad.
            # pylint: disable=unexpected-keyword-arg
            return self.client.get(
                index=searchindex_id, id=event_id,
                _source_exclude=[u'timesketch_label'])
        except NotFoundError:
            abort(HTTP_STATUS_CODE_NOT_FOUND)

**Comment by adamjnichols** [comment]
Python class methods always accept a copy of the class instance as the implied first arg.  Hence "takes at least 4 arguments (4 given)."

Have you run with debug enabled?  What was that output?

**Comment by NetwarSystem** [comment]
I'm not up on object oriented Python, I still write like a Pascal programmer :-(

There are Gists above showing errors with both Elasticsearch 5.6.5 and 6.1.1 - this is what is in /etc/timesketch.conf

# Show debug information.
#
# Note: It is a security risk to have this enabled in production.
DEBUG = True

**Comment by adamjnichols** [comment]
Have you had a look in /var/log/elasticsearch/elasticsearch.log?

**Comment by NetwarSystem** [comment]
I added this to /usr/local/lib/python2.7/dist-packages/timesketch/lib/datastores/elastic.py around line 291:

print(searchindex_id)
print(event_id)

It shows this:

74376fd9e8d748a4bb548c732c58b29f

RX2F4mABb0XgBIcV5SbQ

And this is in execution log right before the trouble:

"GET /api/v1/sketches/2/event/?event_id=RX2F4mABb0XgBIcV5SbQ&searchindex_id=74376fd9e8d748a4bb548c732c58b29f HTTP/1.1" 500 -

Here's the list of indices - three docs in 74376fd9e8d748a4bb548c732c58b29f as expected.

health status index                            uuid                   pri rep docs.count docs.deleted store.size pri.store.size
yellow open   b7668cb644844dcda3a24e95eb573dac gqAa71JKQEuOJN3wzKb0hg   5   1          1            0      7.2kb          7.2kb
yellow open   .kibana                          YVCNFZDqQE-0-azDE96IqQ   5   1        240            2    245.5kb        245.5kb
yellow open   74376fd9e8d748a4bb548c732c58b29f X88k7ujbRPC8KXqLIn7M-A   5   1          3            0     19.3kb         19.3kb

Nothing unusual here, right?

grep "74376fd9e8d748a4bb548c732c58b29f" /var/log/elasticsearch/elasticserch.log

[2018-01-10T18:01:54,048][INFO ][o.e.c.m.MetaDataCreateIndexService] [yb6lxeo] [74376fd9e8d748a4bb548c732c58b29f] creating index, cause [api], templates [], shards [5]/[1], mappings [generic_event]

[2018-01-10T18:01:54,136][INFO ][o.e.c.m.MetaDataMappingService] [yb6lxeo] [74376fd9e8d748a4bb548c732c58b29f/X88k7ujbRPC8KXqLIn7M-A] update_mapping [generic_event]

**Comment by adamjnichols** [comment]
Can you provide the content of elasticsearch.log from around the time you encounter the 500 error please?

**Comment by NetwarSystem** [comment]
I am running Timesketch colocated with Wazuh, wanting to use them together. Let me create a dedicated VM and reproduce this error.

**Comment by NetwarSystem** [comment]
This is /var/log/elasticsearch/elasticsearch.log on a small Ubuntu VM running Elastic 5.6.5.

https://gist.github.com/NetwarSystem/22d973eebff95b5b078b79057b949ae3

And here's the run log for Timesketch

https://gist.github.com/NetwarSystem/8c92a65f0c46be6facebd77f6b449978

**Comment by adamjnichols** [comment]
Thanks for those uploads.  The snippets have a bit of a time gap between them -- are there no further logs on the elasticsearch side at the time you make the request?

This is very odd -- it appears the event_id you're requesting doesn't exist.  Depending on the reason for the last entry in your ES snippet, I might have an idea why, but I'm curious if you have something more in line with the request.

**Comment by NetwarSystem** [comment]
OK, built a new VM from scratch, installed Timesketch, created a three line csv file. Fails to import via the web interface. Import it manually via the command line, same bug as mentioned above. Timestamps now are within a couple minutes, did a full recording of the run with some comments embedded. 

https://gist.github.com/NetwarSystem/b1511b5dac3feb1ff9c1dfa6bba591e1

**Comment by NetwarSystem** [comment]
Same experiment run on Lubuntu 17.10. Prior was on 16.04.3, same failure modes seen here.

https://gist.github.com/NetwarSystem/9af44dc7ee86e5a227d66d45691f50a4

**Comment by NetwarSystem** [comment]
Installed Kibana, three documents, as imported.

https://imgur.com/a/jeSTb

**Comment by berggren** [comment]
This could be an old bug in version 20170721. Can you try to install timesketch from HEAD:
```
git clone https://github.com/google/timesketch.git
cd timesketch
pip install . --upgrade
```

Let me know, I think this issue is fixed since last release.

**Comment by NetwarSystem** [comment]
This is different, but I think not better :-(

https://imgur.com/a/YONTF

Tried another browser, login screen looks right, but still gives this text only stuff after auth.

**Comment by berggren** [comment]
Ah, you need to build the JS/CSS bundle. This is for developers, but if you run HEAD you need to build the latest bundle as well. In the pypi releases we do this for you.

```
  # Add Node.js 8.x repo
  curl -sS https://deb.nodesource.com/gpgkey/nodesource.gpg.key | sudo apt-key add -
  VERSION=node_8.x
  DISTRO="$(lsb_release -s -c)"
  echo "deb https://deb.nodesource.com/$VERSION $DISTRO main" | sudo tee /etc/apt/sources.list.d/nodesource.list

  # Add Yarn repo
  curl -sS https://dl.yarnpkg.com/debian/pubkey.gpg | sudo apt-key add -
  echo "deb https://dl.yarnpkg.com/debian/ stable main" | sudo tee /etc/apt/sources.list.d/yarn.list

cd timesketch
yarn install
yarn build
```

**Comment by NetwarSystem** [comment]
This was close - I would add the following:

apt update
apt install yarn

Otherwise system will suggest installing cmdtest to get yarn, which is not what is needed.

It looks really strange to me that one would install first, then build. Once it's done though ...

https://imgur.com/a/AtRq5

And the detailed log of failures:

https://gist.github.com/NetwarSystem/98a958f84f2ee3ff8434633564925159

**Comment by NetwarSystem** [comment]
I had another go at this, after noticing there had been some commits, still having failures.

Is there an expected date for a new release that is going to fix these things? I'm glad the Vagrant issues are resolved so I can play locally, but I really just want to push this out so half a dozen people can start using it a bit.

**Comment by berggren** [comment]
We are planning on starting releasing more often, but I plan on getting the next release out end of January.

This being said, I will try out the installation docs myself tomorrow and let you know.

**Comment by adamjnichols** [comment]
This issue appears to be related to the mixing of versions and configs.  You seem to have a newer release running, parsing an older version of /etc/timesketch.conf.

Most of the issues I see you're encountering have to do with the complexity of the system you're trying to install, and I think you might be mixing some things up as you work through them out of frustration.  Please e-mail me directly at adamjnichols@gmail.com, and I can guide you more directly offline to ensure you're able to get something going.

**Comment by NetwarSystem** [comment]
OK, I have spent a lot of today on this - I have a reproducible bug for an install of both the current Timesketch via PyPI as well as the github repository under Ubuntu 16.04.3 LTS.

I have a workaround that I've made function on a machine with a public IP and I just duplicated that process on a Lubuntu 16.04.3 LTS instance.

The solution is, roughly:

1) Install per the manual install instructions using pip install timesketch
2) git clone repository, install timesketch as described above
3) install yarn, do yarn update as above
4) pip uninstall timesketch
5) pip install timesketch - so back to the PyPi version

I manually compared the pip freeze output from the working public IP system and the VM I was implementing - there is no difference of any substance in terms of libraries included or versions used. I can offer no theory why the reinstall of timesketch is the final step, but it seems necessary.

I'm used to doctoring Python stuff, but I'm utterly new to yarn, so I don't understand what I've accomplished with the install & build, but I'm delighted I have a machine I can turn over to my guys for further evaluation.

If someone else can confirm this is a problem, I'm happy to turn over the tmux logs of this workaround, I'm probably going to keep grinding on this, maybe see if I can get Timesketch to run in my cluster environment, and if there is some specific testing to be done, just point me the right direction and I'll get going.

I'd like to thank Adam and Johan - hope I don't sound ungrateful with all the bug chatter.
